Separate internal component handling from component implementation

andrewiggins opened the pull request:
My goal with this change is to better understand what parts of our component implementation are generic and which are specific to React-style components. So I've added a separate `renderReactComponent` which is intended to only contain code for rendering react-style  components and `renderComponent` contains generic code for rendering components agnostically. There is a lot of cleanup I'd like to do and I'm not satisfied with some of the existing hacks here, but I thought I'd share to get some early feedback.

This PR is probably better understood by looking at the commits individually. I'll try to summarize the changes here:

* Add `renderReactComponent` function (481fad5)
* Add `SKIP` sentinel to communicate children should be skipped (481fad5) (bleh, currently not a fan of this)
* Add `setGlobalContext` function to allow a component implementation to change the globalContext passed to its children (481fad5) (bleh, currently not a fan of this)
* Always pass newVNode to renderComponent (c4c2fe5)
* Move VNode equality check into patch so we handle both component and DOM node VNode equality in the same place (3666939) (Note: b598683 first moves it outside of `renderReactComponent`)
* Move DIRTY_BIT handling out of component implementation (5e93a1b)
* Copy props from VNode to internal in `renderComponent` so component implementation doesn't have to (b687b49)

The remaining pieces to figure out are to determine how we can do react component things without importing from Preact internals, assuming that's something we are interested in pursuing 😅

andrewiggins committed:
Initial separation of react component from render component

andrewiggins committed:
Always pass newVNode to renderComponent

andrewiggins committed:
Move VNode equality check outside of react component check

andrewiggins committed:
Move VNode equality check into patch

andrewiggins committed:
Move DIRTY_BIT handling out of react component code

andrewiggins committed:
Ensure props are copied over from VNode

andrewiggins committed:
Add globally shared rendererState variable

This commit introduces a rendererState shared "global" variable that contains the state of the renderer we want to share with a component implementation. Currently, it just holds the current context for this component's subtree. Because the globalContext is now managed... well.. globally... we no longer need to pass it down as a parameter to all the functions.

This commit also changes has context is managed. Instead of every component storing its globalContext on itself, only the components that change the context will store a reference to it, and this reference is now stored on Internals. So when we rerender an internal, we climb up the tree to the nearest internal that modified context, stopping at the root, which has context set to an empty object.

andrewiggins committed:
Add skip property to rendererState

This commit replaces the SKIP return signal with a mutable property on rendererState that a component plugin can modify to instruct the diff to skip rendering this component's children (similar to memo or shouldComponentUpdate)

andrewiggins committed:
Add force flag to RendererState

This commit adds a `force` boolean property to renderer state the component plugins must read and ensure that this component is rerendered when this property is true

andrewiggins committed:
Fix linting errors

andrewiggins committed:
Add commit property to RendererState

This change adds a 'commit' property to RendererState to inform the diff if this component internal should be committed, i.e. if this component should be added to the commit queue and have options.commit invoked on it.

andrewiggins committed:
Move react component specific unmounting code to reactComponent file

andrewiggins committed:
Add some helpful TODOs and type comments

andrewiggins committed:
Merge branch 'restructure' into restructure-component-prep

andrewiggins committed:
Add follow ups

andrewiggins commented on follow-ups.md:
```diff
@@ -6,6 +6,18 @@
 
+## Component plugin prep follow ups
+
+- Consider exposing flags to replace various bools in `rendererState`
+	- Move setting COMMIT_FLAG to `setState`/`forceUpdate`
+- Consider adding TYPE_ERROR_BOUNDARY
```
These are already addressed in #3107

marvinhagemeister reviewed (approved):
I like these changes a lot! Not having to drag a `_globalContext` around and instead switching the one from the component is much nicer and requires less pointers to keep around. This is a fantastic refactor 🙌 💯

developit commented:
This all looks good (and like a boatload of work, nice job). One thing that was gnawing at me as I read through though: what was the reason for moving `globalContext` to be global instead of being an object passed down through the tree as a function parameter? I recall we talked about this, but I don't remember the reasoning.

JoviDeCroock commented on src/diff/reactComponents.js:
```diff
@@ -0,0 +1,195 @@
+ * RendererState interface.
+ * @returns {import('../internal').PreactNode} pointer to the next DOM node (in order) to be rendered (or null)
+ */
+export function renderReactComponent(newVNode, internal, rendererState) {
```
Personally would name this `renderComponent` rather than a specific name

JoviDeCroock commented on src/diff/component.js:
```diff
@@ -27,173 +30,41 @@ export function renderComponent(
-			// returns Fragments with multiple DOM children
-			return reorderChildren(internal, startDom, parentDom);
-		}
+	const renderResult = renderReactComponent(newVNode, internal, rendererState);
```
I wonder if it would be worth differentiating and making a hot path: if class `renderComponent` if func `type.render()`

andrewiggins committed:
Expose flags to component plugins (#3107)

* Make `flags` a public field on Internal

* Replace RendererState.force with FORCE_UPDATe flag

* Move RendererState.commit to be a COMMIT_COMPONENT flag

* Replace RendererState.skip with a SKIP_CHILDREN flag

* Remove unneccessary export of createContext counter

* Add TODO to revisit rendererState

* Add TYPE_ERROR_BOUNDARY and extract React error handling code from catchError

This commit removes TYPE_CLASS since our renderer is now abstracted from how components render.

It also removes a test that no longer makes sense since our catch error
doesn't stop at every component and checks that it is dirty, but instead
only stops at error boundaries

* Add notes about other omponent plugin considerations

* Remove out-of-date comment

* Update follow ups

JoviDeCroock commented:
Closing: targets the abandoned `restructure` branch (component-internals split that was not pursued in the shipped v11).
